presto-main/src/main/java/io/prestosql/execution/StateMachine.java

```diff
@@ -83,7 +83,23 @@ public StateMachine(String name, Executor executor, T initialState, Iterable<T>
         this.name = requireNonNull(name, "name is null");
         this.executor = requireNonNull(executor, "executor is null");
         this.state = requireNonNull(initialState, "initialState is null");
-        this.terminalStates = ImmutableSet.copyOf(requireNonNull(terminalStates, "terminalStates is null"));
+        this.isTerminalState = state -> ImmutableSet.of(terminalStates).contains(state);
```

Comment:
Make this ctor delegate to the other ctor.

Also, `ImmutableSet.of(terminalStates)::contains`

Reply:
It's already so at the next commit :)